File: src/routes/events.ts
import { FastifyInstance } from 'fastify';
import { randomUUID } from 'node:crypto';

const EVENT_TYPES = [
  'device.connected',
  'device.disconnected',
  'fabric.transaction',
  'proof.verified',
  'proof.failed',
  'build.completed',
  'agent.started',
  'trace.completed',
  'security.prompt'
] as const;

export default async function eventRoutes(fastify: FastifyInstance) {
  fastify.get(
    '/stream',
    { preHandler: [(fastify as any).authenticate] },
    async (request, reply) => {
      // Establish SSE
      reply.raw.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
        'X-Accel-Buffering': 'no'
      });

      const send = (event: string, data: unknown) => {
        const envelope = {
          id: randomUUID(),
          type: event,
          timestamp: new Date().toISOString(),
          source: 'ortho32-api',
          correlation_id: (request.headers['x-correlation-id'] as string) || randomUUID(),
          data
        };
        reply.raw.write(`event: ${event}\n`);
        reply.raw.write(`data: ${JSON.stringify(envelope)}\n\n`);
      };

      // Initial hello
      send('connected', { message: 'sse connected', events: EVENT_TYPES });

      const interval = setInterval(() => {
        const ev = EVENT_TYPES[Math.floor(Math.random() * EVENT_TYPES.length)];
        const dataByType: Record<string, unknown> = {
          'device.connected': { deviceId: 'ortho0' },
          'device.disconnected': { deviceId: 'sim0' },
          'fabric.transaction': { id: `txn-${Date.now()}`, cycles: 42 },
          'proof.verified': { proofId: randomUUID(), status: 'VERIFIED' },
          'proof.failed': { proofId: randomUUID(), status: 'FAILED' },
          'build.completed': { buildId: randomUUID(), status: 'success' },
          'agent.started': { agentId: 'verify' },
          'trace.completed': { traceId: randomUUID(), cycles: 256 },
          'security.prompt': { prompt: 'elevated_scope_request', scope: 'hardware.reset' }
        };
        send(ev, dataByType[ev] ?? {});
      }, 1000);

      // Heartbeat to keep connection alive
      const heartbeat = setInterval(() => {
        reply.raw.write(`: heartbeat ${Date.now()}\n\n`);
      }, 15000);

      request.raw.on('close', () => {
        clearInterval(interval);
        clearInterval(heartbeat);
        try {
          reply.raw.end();
        } catch {}
      });
    }
  );
}

File: app/routes/events.py
from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import StreamingResponse
import json
import asyncio

from app.ipc import get_host_connection, IPCUnavailable

router = APIRouter(prefix="/events", tags=["events"])

ALLOWED_EVENTS = {
    "device.connected",
    "device.disconnected",
    "fabric.transaction",
    "proof.verified",
    "proof.failed",
    "build.completed",
    "agent.started",
    "trace.completed",
}


@router.get("/stream")
async def stream_events(request: Request):
    host = get_host_connection()

    async def event_generator():
        try:
            # subscribe to all events (wildcard)
            async for env in host.subscribe("*"):
                if await request.is_disconnected():
                    break
                # filter to allowed types if needed; allow all but ensure mapping
                event_type = env.type
                # normalize dot notation: host may send DEVICE_CONNECTED etc, map to allowed
                data = json.dumps(env.model_dump())
                yield f"event: {event_type}\n"
                yield f"data: {data}\n\n"
                await asyncio.sleep(0)
        except IPCUnavailable as e:
            # stream error as SSE comment then close
            yield f"event: error\n"
            yield f"data: {json.dumps({'detail': f'Host unavailable: {str(e)}'})}\n\n"
        except asyncio.CancelledError:
            return

    return StreamingResponse(event_generator(), media_type="text/event-stream", headers={
        "Cache-Control": "no-cache",
        "Connection": "keep-alive",
        "X-Accel-Buffering": "no",
    })
